src/execution_plan/ops/op_merge.c

```diff
@@ -15,193 +14,289 @@
 
-	if(op->stats) op->stats->properties_set += props->property_count;
-}
+	Schema *s = GraphContext_GetSchemaByID(gc, label_id, SCHEMA_NODE);
+	if(!Schema_HasIndices(s)) return; // No indices, no need to update.
 
-/* Saves every entity within the query graph into the actual graph.
- * update statistics regarding the number of entities create and properties set. */
-static void _CommitNodes(OpMerge *op, Record r) {
-	int labelID;
-	Graph *g = op->gc->g;
+	Schema_AddNodeToIndices(s, n, true);
 
-	uint node_count = array_len(op->nodes_to_merge);
+}
 
-	// Start by creating nodes.
-	Graph_AllocateNodes(g, node_count);
+// Update the appropriate property on a graph entity.
+static void _UpdateProperty(Record r, GraphEntity *ge, EntityUpdateEvalCtx *update_ctx) {
+	SIValue new_value = AR_EXP_Evaluate(update_ctx->exp, r);
 
-	for(uint i = 0; i < node_count; i ++) {
-		NodeCreateCtx *node_ctx = &op->nodes_to_merge[i];
-		// Get blueprint of node to create
-		QGNode *node_blueprint = node_ctx->node;
+	// Try to get current property value.
+	SIValue *old_value = GraphEntity_GetProperty(ge, update_ctx->attribute_idx);
 
-		// Newly created node will be placed within given record.
-		Node *created_node = Record_GetNode(r, node_ctx->node_idx);
+	if(old_value == PROPERTY_NOTFOUND) {
+		// Add new property.
+		GraphEntity_AddProperty(ge, update_ctx->attribute_idx, new_value);
+	} else {
+		// Update property.
+		GraphEntity_SetProperty(ge, update_ctx->attribute_idx, new_value);
+	}
+}
 
-		Schema *schema = NULL;
+// Perform all ON MATCH updates for all matched records.
+static void _UpdateProperties(OpMerge *op, Record *records) {
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	Graph_AcquireWriteLock(gc->g); // Lock the graph.
 
-		// Set, create label.
-		if(node_blueprint->label == NULL) {
-			labelID = GRAPH_NO_LABEL;
-		} else {
-			schema = GraphContext_GetSchema(op->gc, node_blueprint->label, SCHEMA_NODE);
-			/* This is the first time we've encountered this label; create its schema */
-			if(schema == NULL) {
-				schema = GraphContext_AddSchema(op->gc, node_blueprint->label, SCHEMA_NODE);
-				op->stats->labels_added++;
-			}
-			labelID = schema->id;
-		}
+	// Iterate over all update contexts, converting property keys to IDs.
+	uint update_count = array_len(op->on_match);
+	for(uint i = 0; i < update_count; i ++) {
+		op->on_match[i].attribute_idx = GraphContext_FindOrAddAttribute(gc, op->on_match[i].attribute);
+	}
+
+	uint record_count = array_len(records);
+	for(uint i = 0; i < record_count; i ++) {  // For each matched record
+		Record r = records[i];
+		for(uint j = 0; j < update_count; j ++) { // For each pending update.
+			EntityUpdateEvalCtx *update_ctx = &op->on_match[j];
 
-		Graph_CreateNode(g, labelID, created_node);
+			// Retrieve the appropriate entry from the Record, make sure it's either a node or an edge.
+			RecordEntryType t = Record_GetType(r, update_ctx->record_idx);
+			assert(t == REC_TYPE_NODE || t == REC_TYPE_EDGE);
```

Comment:
If we do want to perform this check let's do it outside the lock.

Reply:
The Record index can change between different `EntityUpdateEvalCtx` structs:
`MATCH (a:A), (b:B) MERGE (a)-[:E]->(b) ON MATCH SET a.val = 1 ON MATCH SET b.val = 2`
And we want to differentiate between nodes and edges because we might need to update node indices.